File: core/src/main/java/org/datasyslab/geospark/jts/geom/MultiPolygon.java
/*
 * FILE: MultiPolygon
 * Copyright (c) 2015 - 2019 GeoSpark Development Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.datasyslab.geospark.jts.geom;

import org.locationtech.jts.geom.Geometry;

import java.util.Objects;

/**
 * Wraps {@link org.locationtech.jts.geom.MultiPolygon}
 */
public class MultiPolygon extends org.locationtech.jts.geom.MultiPolygon {

    public static org.locationtech.jts.geom.Polygon[] getPolygons(org.locationtech.jts.geom.GeometryCollection geometry) {
        org.locationtech.jts.geom.Polygon[] res = new org.locationtech.jts.geom.Polygon[geometry.getNumGeometries()];
        for (int i = 0; i < res.length; i++)
            res[i] = (org.locationtech.jts.geom.Polygon) geometry.getGeometryN(i);
        return res;
    }

    private static Polygon[] convertPolygons(org.locationtech.jts.geom.Polygon[] polygons) {
        GeometryFactory factory = new GeometryFactory();
        Polygon[] res = new Polygon[polygons.length];
        for (int i = 0; i < res.length; i++) {
            res[i] = (Polygon)factory.fromJTS(polygons[i]);
        }
        return res;
    }

    public MultiPolygon(Object original) {
        this((org.locationtech.jts.geom.MultiPolygon) original);
    }

    public MultiPolygon(org.locationtech.jts.geom.MultiPolygon original) {
        this(getPolygons(original), original.getFactory());
        GeometryCommonUtils.initUserDataFrom(this, original);
    }

    /**
     * {@link org.locationtech.jts.geom.MultiPolygon#MultiPolygon(org.locationtech.jts.geom.Polygon[], org.locationtech.jts.geom.GeometryFactory)}
     */
    public MultiPolygon(org.locationtech.jts.geom.Polygon[] polygons, org.locationtech.jts.geom.GeometryFactory factory) {
        super(convertPolygons(polygons), new GeometryFactory(factory));
        GeometryCommonUtils.initUserDataFrom(this, this);
    }

    /**
     * Compares to given geometry using {@link org.locationtech.jts.geom.MultiPolygon#equals(Geometry)}
     * Also compares userData
     */
    @Override
    public boolean equals(Geometry g) {
        return super.equals(g) && GeometryCommonUtils.userDataEquals(this, g);
    }

    /**
     * Compares to given geometry using {@link org.locationtech.jts.geom.MultiPolygon#equals(Object)}
     * Also compares userData
     */
    @Override
    public boolean equals(Object o) {
        return super.equals(o) && GeometryCommonUtils.userDataEquals(this, o);
    }

    /**
     * Produces {@link MultiPolygon#toString()} (a WKT representation of the geometry)
     *  , concatenated with the userData string (if exists) as a TSV
     */
    @Override
    public String toString() {
        return GeometryCommonUtils.makeString(super.toString(), getUserData());
    }
}

File: python/sedona/core/serde/shape/multipolygon.py
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing,
#  software distributed under the License is distributed on an
#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#  KIND, either express or implied.  See the License for the
#  specific language governing permissions and limitations
#  under the License.

import attr
from shapely.geometry import MultiPolygon

from sedona.core.serde.binary.buffer import BinaryBuffer
from sedona.core.serde.binary.order import ByteOrderType
from sedona.core.serde.binary.parser import BinaryParser
from sedona.core.serde.shape.enums import GeomEnum
from sedona.core.serde.shape.util import get_number_of_polygon_points, get_number_of_rings, add_shape_geometry_metadata, \
    add_offsets_to_polygon, reverse_linear_ring, put_coordinates
from sedona.utils.abstract_parser import GeometryParser


@attr.s
class MultiPolygonParser(GeometryParser):
    name = "MultiPolygon"

    @classmethod
    def serialize(cls, obj: MultiPolygon, binary_buffer: BinaryBuffer):
        if isinstance(obj, MultiPolygon):
            num_polygons = len(obj.geoms)
            num_points = sum([get_number_of_polygon_points(polygon) for polygon in obj.geoms])
            num_rings = sum([get_number_of_rings(polygon) for polygon in obj.geoms])
            add_shape_geometry_metadata(GeomEnum.polygon.value, binary_buffer)
            binary_buffer.add_empty_bytes("double", 4)
            binary_buffer.put_int(num_rings, ByteOrderType.LITTLE_ENDIAN)
            binary_buffer.put_int(num_points, ByteOrderType.LITTLE_ENDIAN)

            offset = 0
            for geom in obj.geoms:
                offset = add_offsets_to_polygon(geom, binary_buffer, offset)

            for geom in obj.geoms:
                coordinates_exterior = reverse_linear_ring(geom.exterior, False)
                put_coordinates(coordinates_exterior, binary_buffer)

                for ring in geom.interiors:
                    coordinates = reverse_linear_ring(ring)
                    put_coordinates(coordinates, binary_buffer)

            binary_buffer.put_byte(1)
            binary_buffer.put_byte(3)
            binary_buffer.put_byte(1)
            binary_buffer.put_byte(-127)
        else:
            raise TypeError(f"Need a {cls.name} instance")
        return binary_buffer.byte_array

    @classmethod
    def deserialize(cls, parser: BinaryParser) -> MultiPolygon:
        raise NotImplementedError("For multipolygon, PolygonParser class is used.")
